File: load-test-framework/ruby_src/cps_subscriber_task.rb
#!/usr/bin/ruby

require 'grpc'
require 'thread'
require 'google/cloud/pubsub'
require './loadtest_services_pb'

include Google::Pubsub::Loadtest

class ServerImpl < Google::Pubsub::Loadtest::LoadtestWorker::Service

  def start(request, _call)
    puts "Received Start"
    @latencies = []
    @recv_msgs = []
    @semaphore = Mutex.new
    pubsub = Google::Cloud::Pubsub.new(project: request.project)
    sub = pubsub.subscription request.pubsub_options.subscription
    @subscriber = sub.listen do |msg|
      puts "Received message " + msg.to_s
      @semaphore.synchronize do
        @latencies.push (Time.now.to_f * 1000.0 - msg.attributes['sendTime'].to_f).to_i
	@recv_msgs.push MessageIdentifier.new(publisher_client_id: msg.attributes['clientId'].to_i, sequence_number: msg.attributes['sequenceNumber'].to_i)
      end
      msg.ack!
    end
    @subscriber.start
    StartResponse.new
  end

  def execute(request, _call)
    latencies = []
    recv_msgs = []
    @semaphore.synchronize do
      latencies = @latencies
      recv_msgs = @recv_msgs
      @latencies = []
      @recv_msgs = []
    end
    puts "Returning " + latencies.to_s + ", " + recv_msgs.to_s
    ExecuteResponse.new(latencies: latencies, received_messages: recv_msgs)
  end
end

def main
  s = GRPC::RpcServer.new
  s.add_http2_port("0.0.0.0:6000", :this_port_is_insecure)
  s.handle(ServerImpl.new)
  puts "Started server on port 6000."
  s.run_till_terminated
end

main

File: load-test-framework/dotnet_src/Subscriber/CPSSubscriberTask.cs
ing Google.Cloud.PubSub.V1;
using Grpc.Core;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;


namespace Google.Pubsub.Loadtest
{
    class LoadtestWorkerImpl : LoadtestWorker.LoadtestWorkerBase
    {
        private static readonly DateTime Jan1st1970 = new DateTime(
            1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
        private Mutex mutex;
        private SimpleSubscriber client;
        private List<long> latencies;
        private List<MessageIdentifier> received_messages;

        public override Task<StartResponse> Start(StartRequest request,
                                                  ServerCallContext context)
        {
            mutex = new Mutex();
            latencies = new List<long>();
            received_messages = new List<MessageIdentifier>();
            client = SimpleSubscriber.CreateAsync(
                new SubscriptionName(request.Project,
                                     request.PubsubOptions.Subscription)
            ).Result;
            client.StartAsync((msg, ct) =>
            {
                long now = CurrentTimeMillis();
                var identifier = new MessageIdentifier();
                mutex.WaitOne();
                latencies.Add(now - long.Parse(msg.Attributes["sendTime"]));
                identifier.PublisherClientId =
                              long.Parse(msg.Attributes["clientId"]);
                identifier.SequenceNumber =
                              int.Parse(msg.Attributes["sequenceNumber"]);
                received_messages.Add(identifier);
                mutex.ReleaseMutex();
                return Task.FromResult(SimpleSubscriber.Reply.Ack);
            });
            return Task.FromResult(new StartResponse());
        }

        public static long CurrentTimeMillis()
        {
            return (long)(DateTime.UtcNow - Jan1st1970).TotalMilliseconds;
        }

        public override Task<ExecuteResponse> Execute(ExecuteRequest request,
                                                      ServerCallContext context)
        {
            ExecuteResponse executeResponse = new ExecuteResponse();
            mutex.WaitOne();
            executeResponse.Latencies.Add(latencies);
            executeResponse.ReceivedMessages.Add(received_messages);
            latencies.Clear();
            received_messages.Clear();
            mutex.ReleaseMutex();
            return Task.FromResult(executeResponse);
        }

        public static void Main(string[] args)
        {
            // Build server
            var server = new Server
            {
                Services =
                {
                    LoadtestWorker.BindService(new LoadtestWorkerImpl())
                },
                Ports =
                {
                    new ServerPort("localhost", 6000,
                                   ServerCredentials.Insecure)
                }
            };

            // Start server
            server.Start();
            Console.WriteLine("Server listening on port 6000.");
            var waiter = new ManualResetEvent(false);
            waiter.WaitOne();
        }
    }
}
